Amazon.QLDB.Driver/driver/QldbDriverBase.cs

```diff
@@ -137,12 +136,41 @@ internal void ReleaseSession(T session)
             }
         }
 
-        internal void LogSessionPoolState()
+        internal void ThrowIfClosed()
+        {
+            if (this.IsClosed)
+            {
+                this.Logger.LogError(ExceptionMessages.DriverClosed);
+                throw new QldbDriverException(ExceptionMessages.DriverClosed);
+            }
+        }
+
+        internal T GetSessionFromPool()
         {
             this.Logger.LogDebug(
                 "Getting session. There are {} free sessions and {} available permits.",
                 this.SessionPool.Count,
                 this.SessionPool.BoundedCapacity - this.PoolPermits.CurrentCount);
+
+            if (this.PoolPermits.Wait(0))
+            {
+                lock (this)
```

Comment:
Rather lock on an internal mutex, not the object itself, which someone else might have locked.

Reply:
Lock removed.